### PR TITLE
Fix broken server-side flow event processing

### LINKED ISSUE #4135: flow.begin event has no metricsContext data
The `flow.begin` event is not emitted with `metricsContext` data. This forces us to do an extra `UPDATE` in the data import script but even that can't help us in the cases where `flow.begin` is the only event in the flow.

### PR DESCRIPTION
**Pull request by philbooth**
Fixes #4135.

In addition to the problem discussed in the linked issue, I found two others that were also breaking the flow event data.

The issues are as follows:

* None of the metrics context fields were being propagated to the event, because the code was mystifyingly trying to copy the data from non-existent query parameters instead of the request payload.

* Flow events were being emitted with their timestamp set to the `flowBeginTime` instead of the actual time of the event.

* Unset values on the metrics context data were being passed in with the default value `"none"`, which conflicts with how they are handled in the auth server.

The fixes for each issue should be quite straightforward but I'll also add some inline commentary in a jiffy.

Sample log lines showing the fixed event data:

```
{"event":"flow.begin","flow_id":"fa5a4b756213e2472e493aa4f3367efe2ba40dd957574ea529b87a448ca7f269","flow_time":0,"hostname":"pbooth-22205.home","op":"flowEvent","pid":10998,"time":"2016-10-04T18:19:00.997Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","v":1,"context":"fx_desktop_v3","entrypoint":"fxa:signup-complete","service":"sync"}
{"event":"flow.signin.engage","flow_id":"fa5a4b756213e2472e493aa4f3367efe2ba40dd957574ea529b87a448ca7f269","flow_time":5662,"hostname":"pbooth-22205.home","op":"flowEvent","pid":10998,"time":"2016-10-04T18:19:06.659Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","v":1,"context":"fx_desktop_v3","entrypoint":"fxa:signup-complete","service":"sync"}
{"event":"flow.signin.submit","flow_id":"fa5a4b756213e2472e493aa4f3367efe2ba40dd957574ea529b87a448ca7f269","flow_time":10382,"hostname":"pbooth-22205.home","op":"flowEvent","pid":10998,"time":"2016-10-04T18:19:11.379Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","v":1,"context":"fx_desktop_v3","entrypoint":"fxa:signup-complete","service":"sync"}
```

@vladikoff, this is the PR I mentioned in IRC earlier, can you give it an r?

If it looks okay, I'll rebase and update our other two flow event PRs tomorrow to try and get everything merged in time for train 71.

### CODE REVIEW

File: server/lib/flow-event.js [hunk 1]
```diff
@@ -6,38 +6,39 @@ var _ = require('lodash');
...
   'utm_source',
   'utm_term'
 ]);
 var HOSTNAME = os.hostname();
-var MAX_PARAM_LENGTH = 100;
+var MAX_DATA_LENGTH = 100;
 var VERSION = 1;
 
 module.exports = function (event, data, request) {
-  var queryParams = _.pick(request.query, isDNT(request) ?
```
Comment (philbooth) [inline server/lib/flow-event.js]: I genuinely have no idea why or how I wrote this to process query params first time round. Maybe I implemented the server stuff first then forgot to come back and check it after the client-side code landed? Bad skills, pb.

File: server/lib/flow-event.js [hunk 2]
```diff
@@ -6,38 +6,39 @@ var _ = require('lodash');
...
     pid: process.pid,
+    time: event.time,
     userAgent: request.headers['user-agent'],
     v: VERSION
-  }, data, _.mapValues(queryParams, limitLength));
+  }, _.mapValues(pickedData, sanitiseData));
 
-  optionallySetFallbackData(eventData, 'service', request.query.client_id);
-  optionallySetFallbackData(eventData, 'entrypoint', request.query.entryPoint);
+  optionallySetFallbackData(eventData, 'service', data.client_id);
```
Comment (philbooth) [inline server/lib/flow-event.js]: I'm not 100% certain whether this `client_id` marshalling will ever kick in. I couldn't make it happen in testing but figured that maybe there is some way for legacy clients to trigger it, so erred on the side of caution.

File: server/lib/flow-event.js [hunk 3]
```diff
@@ -56,12 +57,20 @@ function isDNT (request) {
...
+function limitLength (data) {
+  if (data && data.length > MAX_DATA_LENGTH) {
+    return data.substr(0, MAX_DATA_LENGTH);
+  }
+
+  return data;
+}
+
+function sanitiseData (data) {
+  if (data === 'none') {
```
Comment (philbooth) [inline server/lib/flow-event.js]: Not sure where this default value comes from, presumably elsewhere in the content server. We need to ditch it anyway, for consistency with how the events are emitted in the auth server.

File: server/lib/routes/post-metrics.js [hunk 4]
```diff
@@ -68,31 +68,46 @@ function optionallyLogFlowEvents (req, metrics, requestReceivedTime) {
...
+    } else {
+      event.time = estimateTime({
+        /*eslint-disable sorting/sort-object-props*/
+        start: metrics.startTime,
+        offset: event.offset,
+        sent: metrics.flushTime,
+        received: requestReceivedTime
+        /*eslint-enable sorting/sort-object-props*/
+      });
+      event.flowTime = event.time - metrics.flowBeginTime;
```
Comment (vladikoff) [inline server/lib/routes/post-metrics.js]: 👍

File: server/lib/flow-event.js (hunk 3 above)
Comment (vladikoff) [inline server/lib/flow-event.js]: yea it comes from https://github.com/mozilla/fxa-content-server/blob/b58f42740541ed6ea3cd41f699ab7eb4231a6cf3/app/scripts/lib/metrics.js#L65 probably

**Pull request merged**